pkg/skaffold/kubernetes/context/context.go

```diff
@@ -30,26 +31,44 @@ var (
 )
 
 var (
-	currentConfigOnce sync.Once
-	currentConfig     clientcmdapi.Config
-	currentConfigErr  error
+	kubeConfigOnce sync.Once
+	kubeConfig     clientcmd.ClientConfig
+	kubeContext    string
 )
 
-// ResetCurrentConfig is used by tests
-func ResetCurrentConfig() {
-	currentConfigOnce = sync.Once{}
+// ResetConfig is used by tests
+func ResetConfig() {
+	kubeConfigOnce = sync.Once{}
 }
 
+// UseKubeContext sets an override for the current context in the k8s config.
+// This override must be set before calling CurrentConfig.
+func UseKubeContext(overrideKubeContext string) {
+	kubeContext = overrideKubeContext
+}
+
+func GetRestClientConfig() (*restclient.Config, error) {
+	clientConfig, err := getKubeConfig().ClientConfig()
+	return clientConfig, errors.Wrap(err, "error creating kubeConfig")
+}
+
+// getCurrentConfig retrieves the kubeconfig file. If UseKubeContext was called before, the CurrentContext will be overridden.
 func getCurrentConfig() (clientcmdapi.Config, error) {
-	currentConfigOnce.Do(func() {
```

Comment:
why did we change the semantics of the `currentConfigOnce`? 
Doesn't this open up the possibilities of inconsitencies in a skaffold process that started off version1 of a kubeconfig file but then someone mutated that file and a second read would be of version2?

Reply:
The once-semantic is now encapsulated in `getKubeConfig`. So for the moment we are safe here. However, in the future when the kube-context is also overridden by `skaffold.yaml`, we will have to re-read the kubeconfig. This will occur during the `skaffold.yaml` parsing/profile phase and should not take very long. Do you think Skaffold should take extra precautions to work with the identical kubeconfig as in the first kubeconfig read?

Reply:
I found a nice way to ensure that Skaffold always uses the kubeconfig from the first read. Note that this has not been the case with respect to the kubernetes REST client so far, and is much more consistent now.

Note that there is still the possibility, that the kubecontext changes due to a change in `skaffold.yaml` (via `UseKubeContext`). However, that is the desired behavior.